src/pykeen/nn/pyg.py

```diff
@@ -0,0 +1,391 @@
+        :param layers_kwargs:
+            additional keyword-based parameters passed to the layers upon instantiation
+
+        :param base:
+            the base representations for entities, or a hint thereof
+        :param base_kwargs:
+            additional keyword-based parameters passed to the base representations upon instantiation
+
+        :param output_shape:
+            the output shape. Defaults to the base representation shape. Has to match to output shape of the last
+            message passing layer.
+
+        :param activation:
+            the activation(s), or hints thereof
+        :param activation_kwargs:
+            additional keyword-based parameters passed to the activations upon instantiation
+
+        :param kwargs:
+            additional keyword-based parameters passed to :meth:`Representation.__init__`
+
+        :raises ImportError:
+            if PyTorch Geometric is not installed
+        :raises ValueError:
+            if the number of activations and message passing layers do not match (after input normalization)
+        """
+        # fail if dependencies are missing
+        if MessagePassing is None or layer_resolver is None:
+            raise ImportError(_PYG_INSTALLATION_TEXT)
+
+        # avoid cyclic import
+        from . import representation_resolver
+
+        # the base representations, e.g., entity embeddings or features
+        base = representation_resolver.make(base, pos_kwargs=base_kwargs, max_id=triples_factory.num_entities)
+
+        super().__init__(max_id=kwargs.pop("max_id", base.max_id), shape=output_shape or base.shape, **kwargs)
+
+        # assign sub-module *after* super call
+        self.base = base
+
+        # initialize layers
+        self.layers = nn.ModuleList(layer_resolver.make_many(layers, layers_kwargs))
+
+        # normalize activation
+        activation = list(upgrade_to_sequence(activation))
+        if len(activation) == 1:
+            activation = activation * len(self.layers)
+        self.activations = nn.ModuleList(activation_resolver.make_many(activation, activation_kwargs))
+        if len(self.layers) != len(self.activations):
+            raise ValueError(
+                f"The lengths of the list of message passing layers ({len(self.layers)}) "
+                f"and activation layers ({len(self.activation)}) differs! To disable activations "
+                f"on certain layers, e.g., the last, use torch.nn.Identity."
+            )
+
+        # prepare buffer
+        # TODO: inductive?
+        self.register_buffer(name="edge_index", tensor=triples_factory.mapped_triples[:, [0, 2]].t())
+
+    def _plain_forward(self, indices: Optional[torch.LongTensor] = None) -> torch.FloatTensor:  # noqa: D102
+        # TODO: we could reduce the memory footprint and maybe also computation time
+        #       by considering only the k-hop neighborhood of the requested indices
+        #       computing this may be computationally expensive, too
```

Comment:
PyG has a special function [k-hop subgraph](https://pytorch-geometric.readthedocs.io/en/1.6.1/modules/utils.html?highlight=k-hop%20neighbor#torch_geometric.utils.k_hop_subgraph) precisely for that 😉

Reply:
👍 Do you know about the computational overhead of calculating the $k$-hop subgraph, and the still necessary index select at the end?

Reply:
c9a6fe7 - made if optional for now

Reply:
The procedure looks like that of linear complexity. Of course, k-hop neighborhood size grows exponentially with k, but we are not doing subsampling for batches here, we are only shrinking the edge index of the whole graph